File: static/wallet-manager.js
/**
 * Manages wallet connections and transactions.
 */
class WalletManager {
    constructor(config) {
        this.config = config;
        this.wallet = null;
        this.initialized = false;
        this.logger = new Logger('WalletManager');
        this.autoConnect = config.wallet?.auto_connect || false;
    }

    /**
     * Initialize wallet connection.
     */
    async initialize() {
        try {
            this.logger.info('Initializing wallet manager...');

            // Check if Phantom is installed
            if (!window.solana?.isPhantom) {
                throw new Error('Phantom wallet not installed');
            }

            // Try auto-connect if enabled
            if (this.autoConnect) {
                try {
                    await this.connect(true);
                } catch (error) {
                    this.logger.warn('Auto-connect failed:', error.message);
                }
            }

            this.initialized = true;
            this.logger.info('Wallet manager initialized');
            return true;

        } catch (error) {
            this.logger.error('Failed to initialize wallet manager:', error);
            throw error;
        }
    }

    /**
     * Connect to wallet.
     */
    async connect(silent = false) {
        try {
            let resp;
            if (silent) {
                try {
                    resp = await window.solana.connect({ onlyIfTrusted: true });
                } catch {
                    return false;
                }
            } else {
                resp = await window.solana.connect();
            }

            this.wallet = resp.publicKey;
            this.logger.info('Connected to wallet:', this.wallet.toString());

            // Update config with wallet address
            this.config.wallet.address = this.wallet.toString();

            return true;

        } catch (error) {
            this.logger.error('Failed to connect wallet:', error);
            throw error;
        }
    }

    /**
     * Disconnect from wallet.
     */
    async disconnect() {
        try {
            if (window.solana?.isConnected) {
                await window.solana.disconnect();
            }
            this.wallet = null;
            this.logger.info('Disconnected from wallet');

        } catch (error) {
            this.logger.error('Failed to disconnect wallet:', error);
            throw error;
        }
    }

    /**
     * Check if wallet is connected.
     */
    isConnected() {
        return this.wallet !== null && window.solana?.isConnected;
    }

    /**
     * Get wallet address.
     */
    getAddress() {
        return this.wallet?.toString() || null;
    }

    /**
     * Get wallet balance.
     */
    async getBalance() {
        try {
            if (!this.isConnected()) {
                throw new Error('Wallet not connected');
            }

            const connection = await this.getConnection();
            const balance = await connection.getBalance(this.wallet);
            return balance / 1e9; // Convert lamports to SOL

        } catch (error) {
            this.logger.error('Failed to get wallet balance:', error);
            throw error;
        }
    }

    /**
     * Sign and send transaction.
     */
    async signAndSendTransaction(transaction) {
        try {
            if (!this.isConnected()) {
                throw new Error('Wallet not connected');
            }

            // Sign transaction
            const signed = await window.solana.signTransaction(transaction);

            // Send transaction
            const connection = await this.getConnection();
            const signature = await connection.sendRawTransaction(
                signed.serialize(),
                {
                    skipPreflight: false,
                    preflightCommitment: 'confirmed'
                }
            );

            // Confirm transaction
            const confirmation = await connection.confirmTransaction(signature);
            if (confirmation.value.err) {
                throw new Error(`Transaction failed: ${confirmation.value.err}`);
            }

            this.logger.info('Transaction confirmed:', signature);
            return signature;

        } catch (error) {
            this.logger.error('Failed to sign and send transaction:', error);
            throw error;
        }
    }

    /**
     * Close wallet manager.
     */
    async close() {
        try {
            if (this.isConnected()) {
                await this.disconnect();
            }
            this.initialized = false;
            this.logger.info('Wallet manager closed');

        } catch (error) {
            this.logger.error('Failed to close wallet manager:', error);
            throw error;
        }
    }

    /**
     * Get RPC connection.
     */
    async getConnection() {
        if (!window.rpcManager) {
            throw new Error('RPC Manager not initialized');
        }
        return window.rpcManager.getCurrentConnection();
    }
}

// Export for Node.js and browser
if (typeof module !== 'undefined' && module.exports) {
    module.exports = WalletManager;
} else {
    window.WalletManager = WalletManager;
}

File: bot/wallet_manager.py
"""Wallet manager for handling Solana wallets and transactions."""

import logging
import os
from typing import Dict, Any, Optional
from solders.keypair import Keypair
from solders.pubkey import Pubkey
import solders.system_program as system_program
from solders.transaction import Transaction

logger = logging.getLogger(__name__)

class WalletManager:
    """Manages Solana wallets and transactions."""
    
    def __init__(self, config: Dict[str, Any]):
        """Initialize wallet manager."""
        self.config = config
        self.keypair: Optional[Keypair] = None
        self.initialized = False
        self._connected = False
        
    async def initialize(self, rpc_manager):
        """Initialize wallet manager."""
        try:
            if self.initialized:
                return
                
            logger.info("Initializing wallet manager...")
            
            # Store RPC manager reference
            self.rpc = rpc_manager
            
            # Load or create keypair
            await self._load_keypair()
            
            # Test connection
            await self._test_connection()
            
            self.initialized = True
            logger.info("Wallet manager initialized successfully")
            
        except Exception as e:
            logger.error(f"Failed to initialize wallet manager: {str(e)}")
            await self.cleanup()
            raise
            
    async def _load_keypair(self):
        """Load or create Solana keypair."""
        try:
            # Check for existing keypair file
            keypair_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'config', 'keypair.json')
            
            if os.path.exists(keypair_path):
                logger.info("Loading existing keypair...")
                with open(keypair_path, 'r') as f:
                    import json
                    keypair_bytes = bytes(json.load(f))
                    self.keypair = Keypair.from_bytes(keypair_bytes)
            else:
                logger.info("Generating new keypair...")
                self.keypair = Keypair()
                # Save keypair for future use
                os.makedirs(os.path.dirname(keypair_path), exist_ok=True)
                with open(keypair_path, 'w') as f:
                    import json
                    json.dump(list(self.keypair.secret()), f)
                
            logger.info(f"Wallet public key: {self.keypair.pubkey()}")
            
        except Exception as e:
            logger.error(f"Failed to load/create keypair: {str(e)}")
            raise
            
    async def _test_connection(self):
        """Test wallet connection."""
        try:
            if not self.keypair:
                raise Exception("Keypair not initialized")
                
            # Get account info
            response = await self.rpc.client.get_account_info(self.keypair.pubkey())
            if response.value:
                logger.info(f"Account exists with {response.value.lamports / 1e9} SOL")
            else:
                logger.info("Account does not exist yet (needs funding)")
                
            self._connected = True
            
        except Exception as e:
            logger.error(f"Wallet connection test failed: {str(e)}")
            self._connected = False
            raise
            
    async def cleanup(self):
        """Clean up resources."""
        self.initialized = False
        self._connected = False
        
    def is_connected(self) -> bool:
        """Check if wallet is connected."""
        return self._connected
        
    async def get_balance(self) -> float:
        """Get wallet balance in SOL."""
        try:
            if not self.keypair:
                raise Exception("Wallet not initialized")
                
            response = await self.rpc.client.get_balance(self.keypair.pubkey())
            if not response.value:
                return 0.0
            return response.value / 1e9  # Convert lamports to SOL
            
        except Exception as e:
            logger.error(f"Failed to get balance: {str(e)}")
            return 0.0
            
    async def get_token_balance(self, token_mint: str) -> float:
        """Get token balance for a specific SPL token."""
        try:
            if not self.keypair:
                raise Exception("Wallet not initialized")
                
            # Get token account
            response = await self.rpc.client.get_token_accounts_by_owner(
                self.keypair.pubkey(),
                {"mint": token_mint}
            )
            
            if not response.value:
                return 0.0
                
            # Get balance of first token account
            token_account = response.value[0].pubkey
            balance = await self.rpc.client.get_token_account_balance(token_account)
            
            if not balance.value:
                return 0.0
                
            return float(balance.value.amount)
            
        except Exception as e:
            logger.error(f"Failed to get token balance: {str(e)}")
            return 0.0
            
    async def sign_transaction(self, transaction: Transaction) -> Transaction:
        """Sign a transaction with the wallet keypair."""
        try:
            if not self.keypair:
                raise Exception("Wallet not initialized")
                
            transaction.sign([self.keypair])
            return transaction
            
        except Exception as e:
            logger.error(f"Failed to sign transaction: {str(e)}")
            raise
            
    async def send_transaction(self, transaction: Transaction) -> str:
        """Sign and send a transaction."""
        try:
            if not self.keypair:
                raise Exception("Wallet not initialized")
                
            # Sign transaction
            transaction = await self.sign_transaction(transaction)
            
            # Send transaction
            response = await self.rpc.client.send_transaction(
                transaction,
                [self.keypair]
            )
            
            if not response.value:
                raise Exception("Failed to send transaction")
                
            return response.value
            
        except Exception as e:
            logger.error(f"Failed to send transaction: {str(e)}")
            raise
